pTools/mopt.6
.\" mopt.6 by Stefan Strack
.\" $Id: mopt.6,v 1.0 1994/05/21 21:39:48 stst Exp stst $
.\"
.de EX      \"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +5n
..
.de EE      \"End example
.fi
.in -5n
.if n .sp 1
.if t .sp .5
..
.TH MOPT 6 "May 15, 1994" "PMARS PROJECT"
.ad b
.SH NAME
mopt \- multiple constant optimizer
.SH SYNOPSIS
.I mopt
.SH DESCRIPTION
.I Mopt
finds "optimal" combinations of scanning or bombing constants for corewar
warriors. In contrast to other popular optimizers (optima, corestep)
which optimize for a single step value,
.I mopt
is useful for optimizing complex bombing/scanning patterns of stones and
cmp-scanners. The program asks for core size, target size, and up to 10 increment/offset
pairs.
"Increment" is the distance between two consecutive bombs/scans;
"offset" is the position of the first bomb/scan relative to zero.
Increments and offsets can be simply numbers or "generators" that provide a
stream of numbers. If at least one generator is specified,
.I mopt
will find the 20 best combinations of increment/offset pairs, otherwise it
will report the score for the specified increment/offset combination.
Generators have a syntax similar to for-loops in the C language
(see examples below) and use single letter variables A through Z. You don't
have to use them in sequence, i.e. A for increment generator #1, B for offset
generator #2 etc..
.SH EXAMPLES
1) Suppose we have a stone:
.EX
        spl #A,<B
        add -1,1
        mov <-2,-2
        jmp -2
.EE
and we want to find optimal values for increments A and B:
.EX
Coresize: 8000 <enter>
Increment value or generator #1: a=4,a<8000,a=a+4 <enter>
   Offset value or generator #2: 0 <enter>
Increment value or generator #2: b=1,b<8000,b=b+2 <enter>
   Offset value or generator #3: <enter>
.EE
2) Now a carpet-bombing cmp-scanner:
.EX
        sub incr,1
        cmp 0,C
        ...
incr    spl #-A,<-B
        ...

Coresize: 8000 <enter>
Increment value or generator #1: a=2,a<4000,a=a+4 <enter>
   Offset value or generator #2: c=10,c<15,c=c+1 <enter>
Increment value or generator #2: b=a <enter>
   Offset value or generator #3: <enter>
.EE
3) Lastly, a spl/jmp bombing cmp-scanner:
.EX
        sub incr,1
        cmp 0,C      ;C is A/2
        ...
incr    spl #-A,<-B
        ...

Coresize: 8000 <enter>
Increment value or generator #1: a=2,a<4000,a=a+4 <enter>
   Offset value or generator #2: c=a/2 <enter>
Increment value or generator #2: b=a <enter>
   Offset value or generator #3: <enter>
.EE
.SH ALGORITHM
Scoring is based on Nandor Sieben's
iterative optima algorithm with two modifications:
A variable gap cut-off ("target size") is used
and early gaps weigh more than late gaps.
.SH AUTHOR
Stefan Strack (stst@vuse.vanderbilt.edu)


Changes.txt
11/07/2015 - nMars0.9.6 
- moved to github
- Upgraded to VS 2015
- Upgraded to .NET 4.0
- dropped Python module

04/19/2007 - nMars0.9.4 - IDE is running on Linux - Mono 1.2.4
  - still lot of issues on Mono, but we are able to run
  
04/19/2007 - nMars0.9.2 - bugfix release
04/05/2007 - corewars renamed to Core Wars (as requested by William R. Buckley)
03/05/2007 - Fngine, testing, tournament
  - loading always positive numbers to core
  - disable P-space command line option
  - koth format command line option - first step, just option
  - tournament, first vs others engine mode draft
  
02/10/2007 - nMars0.9.1 - bugfix release
  - tooltip fixed
  - active instruction marked red in warrior overview
  - location for more parser errors (expressions)
  - autosave before run
  - quit without save dialog
  - bug in debug warrior overview repainting

02/10/2007 - nMars0.9.0
  - new warrior template
  - breakpoint on doubleclick to cell
  - initialization of round moved to start next step for IDE

02/06/2007 - merged IDE and IDE.Debugger components

02/06/2007 - nMars0.8.4 - tag version before merge IDE and IDE.Debugger components

02/04/2007 - step-back fixes
  - unit test step-back feature
  - step-back fixed, pspace history

01/15/2007 - parser improovements
  - all before ;redcode as comment
  - asserts
  - for 0 comment rof
  - all after end ignore
  - mode as part of equ allowed
  - org expression
  - + sign
  - pMars.DllWrapper fixed error messages
  
01/13/2007 - nMars0.8.3 - IDE split
  - IDE was wplit to IDE, graphics, debuger, debugger.ide
  - MONO proof of concept - result: posible, lot of problems with WinForms, screenshot of debugging
  - Shells and debuggers
  
06/01/2007 - nMars0.8.2 - Refactoring
  - IProject is vehicle for IParser as well for IEngine 
  - Project now contains WarriorFiles and can be saved/loaded to .nmproj , also from pmars config files
  - nMars.exe is new commandline replacement, some new options, hooks for plugins added, configurable components
  - Parser and Engine are no more executables
  - signed assemblies
  - Async Engine moved to RedCode assembly
  - IDE is now showing tootip info about each cell, including last events
  - console debugger is broken now
  - nMars.Engine bugs in instructions fixed
  - Unit test is producing .nmproj setup for failed tests and can be drive by commandline
  - executed 100k matches on combinations of 5k different warriors to test Engine compatibility with pMars. No open issue.
  - started work on documentation comments in code
  - Parser is now returning result structure with list messages, including positions in files
  - pMars.DllWrapper parser is converting error to result structure too, + memory leaks fixed, WARRIORS count is set from rules
  - pMars.dll a-z registry cleanup in dll mode, long filenames buffer overflow fixed
  - class ComponentSetup renamed to ComponentLoader and simplified

12/23/2006 - nMars0.8.1 - IDE
  - IDE framework
  - fast GUI debug watches
  - async engine layer
  - project, solution
  - IDE editor
  - engine layers and iterfaces extended

11/19/2006 - nMars0.8 - debugger
  - console debugger
  - IronPython extension for debugger
  - fixed bugs in engine
  - building step back infrastructure
  - building coredump infrastructure
  
11/13/2006 - nMars0.7 - going public
  - documentation
  - exception handling
  - XML dump
  
11/13/2006 - nMars0.6 - engine unit test
  - refactored engine
  - pMars.dll silent
  - enhanced unit test
    - command line
    - comparer engine
  - dump match results
  - droped module dependencies
  - fixed bugs in engine
     - PSpace support
     - adresssing bugs in engine

11/08/2006 - nMars0.5 - more binaries
  - simplified modules
  - modules loading by name
  - engine commandline
  - IProject introduced
  - version info in binaries
  - release tools in svn
  
11/03/2006 - nMars0.4 - pMars included
  - included pMars circa 0.9.2, with SDL display from M Joonas Pihlaja
  - modularized pMars and wrapped into Windows DLL
  - used pMars.dll as validator for unit tests
  - enhanced SimpleEngine compatiblity, fixed bugs
  - extended interfaces in namespace RedCode

09/28/2006 - nMars0.3 - first binary released
  - Parser more pmarsv compatible
  - mono supported
  - comments parsed
  - unit tests
  - pmarsv wrapper
  
09/17/2006 - nMars0.2 - Parser more powerfull
  - more expression operations
  - label can be A B I ... <Modifier>
  - dynamic label names name & order
  - locations to pareser errors
  - 2 pass label&variable evaluation
  - unit test skeleton
  - exaple warriors in directory rc

08/12/2006 - nMars0.1 - First version

pTools/mopt.c
#define ID "mopt v1.2 - multiple constant optimizer for corewar (c) 1994-96 Stefan Strack\n"
/* $Log: mopt.c,v $
 * Revision 1.3  1996/06/04  07:33:15  stst
 * target size option
 *
 * Revision 1.1  1995/05/27  15:45:42  stst
 * BCC version now with large core support
 *
 * Revision 1.0  1994/05/21  21:42:40  stst
 * Initial revision
 * */
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#if defined(__MSDOS__) && !defined(__GO32__)
#include <alloc.h>
#include <dos.h>
#include <mem.h>
#define MALLOC (char far *) farmalloc
#define FAR far
#define MEMSET _fmemset
#else
#include <malloc.h>
#include <signal.h>
#define MALLOC malloc
#define FAR
#define MEMSET memset
#endif

#define MAXLINE 128
#define MAXCONS 10
#define NBEST 20
#define FOLD(n) (((((n)) % coreSize) + coreSize) % coreSize)
#define E(str,res) (!eval_expr(str,&res) ? res : \
    (fprintf(stderr,"Expression error: \"%s\"\n",str), exit(1), 0))
#define TERMINAL(s) (((s)=='\n') || ((s)=='\r') || ((s)==0))
#define NOEOL(s) do {\
    register char *ch=(s);\
    while (*ch && (*ch != '\n') && (*ch != '\r')) ++ch;\
    *ch=0; } while(0)

typedef long ADDR_T;

typedef struct {
  ADDR_T  ptr;
  struct {
    char    start[MAXLINE], stop[MAXLINE], step[MAXLINE];
    ADDR_T  cur, norm;
  }       offs, incr;
}       consStruct;
typedef struct {
  long    score;
  struct {
    ADDR_T  offs, incr;
  }       cons[MAXCONS];
}       bestStruct;


consStruct cons[MAXCONS];
bestStruct best[NBEST];
int     consNum;
ADDR_T  coreSize;
long    TargetSize;                /* used to be a constant */
char FAR *core;
char   *defCoreSize = "8000";
char   *defTargetSize = "100";

/* prototypes */
#if defined(__STDC__)  ||  defined(__MSDOS__)
long    mopt(void);
void    iterate(int consP);
void    print_best(void);
void    sighandler(int dummy);
extern int eval_expr(char *expr, long *result);
#else
long    mopt();
void    iterate();
void    print_best();
void    sighandler();
extern int eval_expr();
#endif

long
mopt()
{
  ADDR_T  i, j, idx, counter = 0, dist;
  long    accum = 0;
  int     consP = 0;

  for (i = 0; i < consNum; ++i) {
    if (!cons[i].incr.cur)        /* this is legal, e.g. -100,100,5 */
      return 0L;
    cons[i].offs.norm = cons[i].ptr = FOLD(cons[i].offs.cur);
  }
  /* clear core array */
  MEMSET(core, 0, coreSize);

  while (1) {
    if (!consP)                        /* to prevent asymmetries, all bombs hit
                                 * simultaneously */
      for (i = 0; i < consNum; ++i)
        if (!core[cons[i].ptr])
          core[cons[i].ptr] = 1;
    idx = cons[consP].ptr;
    if (core[idx] == 1) {
      core[idx] = 2;
      for (i = (idx - 1 + coreSize) % coreSize, j = (idx + 1) % coreSize, dist = 0;        /* search only until
                                                                                         * TargetSize */
           !core[i] && !core[j] && dist <= TargetSize; --i, ++j, ++dist) {
        if (!i)
          i = coreSize;
        if (j == coreSize - 1)
          j = -1;
      }
      accum += dist * ((coreSize - counter) >> 8);        /* score with decreasing
                                                         * gap weight */
    }
    cons[consP].ptr = cons[consP].ptr + cons[consP].incr.cur;
    while (cons[consP].ptr < 0)        /* faster than FOLD */
      cons[consP].ptr += coreSize;
    while (cons[consP].ptr >= coreSize)
      cons[consP].ptr -= coreSize;
    ++counter;
    if (cons[consP].ptr == cons[consP].offs.norm)
      return accum;
    if (++consP == consNum)
      consP = 0;
  }
}


main()
{
  int     idx, sc1, sc2;
  char    buf1[MAXLINE], buf2[MAXLINE];

  printf(ID);
  do {
    printf("Core size: [%s] ", defCoreSize);
    fgets(buf1, MAXLINE, stdin);
    NOEOL(buf1);
    idx = 0;
    while (isspace(buf1[idx]))
      ++idx;
    if (TERMINAL(buf1[idx]))
      strcpy(buf1, defCoreSize);
  } while (sscanf(buf1, "%ld", &coreSize) != 1 || coreSize <= 0);
  do {
    printf("Target size: [%s] ", defTargetSize);
    fgets(buf1, MAXLINE, stdin);
    NOEOL(buf1);
    idx = 0;
    while (isspace(buf1[idx]))
      ++idx;
    if (TERMINAL(buf1[idx]))
      strcpy(buf1, defTargetSize);
  } while (sscanf(buf1, "%ld", &TargetSize) != 1 || TargetSize <= 0);
  idx = 1;
  while (idx <= MAXCONS) {
    if (idx > 1) {
      printf("   Offset value or generator #%d: ", idx);
      if (!fgets(buf1, MAXLINE, stdin) || buf1[0] <= 13)
        break;
    } else
      strcpy(buf1, "0");
    printf("Increment value or generator #%d: ", idx);
    if (!fgets(buf2, MAXLINE, stdin) || buf2[0] <= 13)
      break;
    if (((sc1 = sscanf(buf1, "%[^,;]%*[,;]%[^,;]%*[,;]%[^,;]", cons[idx - 1].offs.start,
                       cons[idx - 1].offs.stop,
                       cons[idx - 1].offs.step)) == 1 ||
         sc1 == 3) &&
        ((sc2 = sscanf(buf2, "%[^,;]%*[,;]%[^,;]%*[,;]%[^,;]", cons[idx - 1].incr.start,
                       cons[idx - 1].incr.stop,
                       cons[idx - 1].incr.step)) == 1 ||
         sc2 == 3)) {
      if (sc1 == 1) {
        strcpy(cons[idx - 1].offs.stop, "0");
        strcpy(cons[idx - 1].offs.step, "0");
      }
      if (sc2 == 1) {
        strcpy(cons[idx - 1].incr.stop, "0");
        strcpy(cons[idx - 1].incr.step, "0");
      }
      ++idx;
    } else
      printf("Bad format. Examples:\n\t3\n\tb=a/2\n\tc=2,c<100,c=c+2\n");
  }                                /* while(1) */
  consNum = idx - 1;

  if ((core = MALLOC(coreSize)) == NULL) {
    fprintf(stderr, "Cannot allocate core memory\n");
    return 1;
  }
#if  defined(unix) && !defined(__GO32__)
#ifdef SIGINT
  signal(SIGINT, sighandler);
#endif
#else
#if defined(__MSDOS__) && !defined(__GO32__)
  ctrlbrk((int (*) (void)) sighandler);
#endif
#endif

  printf("Hit ^C to see the best so far\n");

  iterate(0);
  printf("The best:\n");
  print_best();
  return 0;
}

void
iterate(consP)
  int     consP;
{
  int     cmp, i;
  long    temp, dummy;

  if (consP == consNum) {
    if ((temp = mopt()) > best[NBEST - 1].score) {        /* sort into top 20 */
      for (cmp = NBEST - 2; cmp >= 0 && temp > best[cmp].score; --cmp) {
        best[cmp + 1].score = best[cmp].score;
        for (i = 0; i < consNum; ++i) {
          best[cmp + 1].cons[i].offs = best[cmp].cons[i].offs;
          best[cmp + 1].cons[i].incr = best[cmp].cons[i].incr;
        }
      }
      best[cmp + 1].cons[0].offs = cons[0].offs.cur;
      printf("%-5ld ", best[cmp + 1].cons[0].incr = cons[0].incr.cur);
      for (i = 1; i < consNum; ++i)
        printf("%-5ld %-5ld ", best[cmp + 1].cons[i].offs = cons[i].offs.cur,
               best[cmp + 1].cons[i].incr = cons[i].incr.cur);
      printf(" -> %f\n", (float) (best[cmp + 1].score = temp) / coreSize);
      fflush(stdout);
    }
  } else {                        /* iterate */
    E(cons[consP].offs.start, cons[consP].offs.cur);
    do {
      E(cons[consP].incr.start, cons[consP].incr.cur);
      do {
        iterate(consP + 1);
        E(cons[consP].incr.step, cons[consP].incr.cur);
      } while (E(cons[consP].incr.stop, dummy));
      E(cons[consP].offs.step, cons[consP].offs.cur);
    } while (E(cons[consP].offs.stop, dummy));
  }                                /* else */
}

void
print_best()
{
  int     i, j;
  for (j = 0; j < NBEST; ++j)
    if (best[j].score > 0) {
      printf("%-5ld ", best[j].cons[0].incr);
      for (i = 1; i < consNum; ++i)
        printf("%-5ld %-5ld ", best[j].cons[i].offs,
               best[j].cons[i].incr);
      printf(" -> %f\n", (float) best[j].score / coreSize);
    }
}

void
sighandler(dummy)
  int     dummy;
{
  char    buf[11];

  printf("The best so far:\n");
  print_best();
  printf("(c)ontinue or (q)uit? ");
  fgets(buf, 10, stdin);
  if (tolower(buf[0]) == 'q')
    exit(dummy);
  else
    return;
}
